src/rust/engine/ui/src/display.rs

```diff
@@ -0,0 +1,307 @@
+      Console::Pipe(ref mut p) => p.write(msg.as_bytes()),
+    }
+  }
+
+  // Renders a divider between the logs and action output.
+  fn render_divider(&mut self, offset: u16) {
+    let cursor_start = self.cursor_start;
+    let padding = self.padding.clone();
+    let divider = self.divider.clone();
+
+    self
+      .write(format!(
+        "{pos}{clear_line}{padding}{blue}{divider}{reset}",
+        pos = cursor::Goto(1, cursor_start.1 + offset),
+        clear_line = clear::CurrentLine,
+        padding = padding,
+        blue = color::Fg(color::Blue),
+        divider = divider,
+        reset = color::Fg(color::Reset)
+      ))
+      .expect("could not write to terminal");
+  }
+
+  // Renders one frame of the log portion of the screen.
+  fn render_logs(&mut self, max_entries: usize) -> usize {
+    let cursor_start = self.cursor_start;
+    let printable_logs: Vec<String> = self.logs.clone().into_iter().take(max_entries).collect();
+
+    let mut counter: usize = 0;
+    for (n, log_entry) in printable_logs.iter().rev().enumerate() {
+      counter += 1;
+      let line_shortened_log_entry: String = format!(
+        "{padding}{log_entry}",
+        padding = self.padding,
+        log_entry = log_entry
+      ).graphemes(true)
+        .take(self.terminal_size.0 as usize)
+        .collect();
+
+      self
+        .write(format!(
+          "{pos}{clear_line}{entry}",
+          pos = cursor::Goto(1, cursor_start.1 + n as u16),
+          clear_line = clear::CurrentLine,
+          entry = line_shortened_log_entry
+        ))
+        .expect("could not write to terminal");
+    }
+
+    if counter > 0 {
+      self.render_divider(counter as u16);
+      counter += 1;
+    }
+    counter
+  }
+
+  // Renders one frame of the action portion of the screen.
+  fn render_actions(&mut self, start_row: usize) {
+    let mut worker_states: Vec<(String, String)> = self.action_map.clone().into_iter().collect();
+
+    worker_states.sort_by_key(|k| k.clone().0);
+
+    let cursor_start = self.cursor_start;
+
+    // For every active worker in the action map, jump to the exact cursor
+    // representing the swimlane for this worker and lay down a text label.
+    for (n, (_worker_id, action)) in worker_states.iter().enumerate() {
+      let line_shortened_output: String = format!(
+          "{padding}{blue}{sigil}{reset}{action}",
+          padding=self.padding,
+          blue=color::Fg(color::LightBlue),
+          sigil=self.sigil,
+          reset=color::Fg(color::Reset),
+          action=action
+        )
+        .graphemes(true)
+        // Account for control characters.
+        .take(self.terminal_size.0 as usize + 14)
+        .collect();
+
+      self
+        .write(format!(
+          "{pos}{entry}",
+          pos = cursor::Goto(1, cursor_start.1 + start_row as u16 + n as u16),
+          entry = line_shortened_output
+        ))
+        .expect("could not write to terminal");
+    }
+  }
+
+  fn render_for_pipe(&self) {
```

Comment:
This one could definitely use some more fleshing out. I think I had been thinking that in the absence of a TTY we would not launch this UI at all, and simply render log messages. Perhaps there is an advantage to having this codepath involved in both modes...?

Reply:
yeah, I was originally thinking that we would also not launch this UI for pipe mode - but in terms of "concerns", having the EngineDisplay handle all rendering seems more ideal than having this + a sidecar pipe impl, I think. handling failure modes also becomes way easier - can't pop the terminal into raw mode for some reason? fallback to pipe mode, etc.

but would be totally open to going either way.

Reply:
Fine with me.